### ISSUE TITLE
RPC cancellation doesn't work if net.Conn.Write blocks

### ISSUE THREAD

**Issue by akalin-keybase** [keybase/go-framed-msgpack-rpc#26]
In `writerLoop()` in transport.go, the call to `t.cdec.Write()` blocks, which causes calls to `d.writer.Encode` in `handleCall` to block, even if `c.ctx` is cancelled. I think we want cancellation to always cause the RPC call to return, per #18.

There are a few complications:
- `writerLoop` still has to preserve FIFO-ness, so even if an RPC call times out, we still need to finish writing its data, since we want messages to arrive whole.
- When an RPC is cancelled, we still have to call `d.writer.Encode` again, to write the MethodCancel error! This also needs to not block.

If you want a test case, see https://github.com/keybase/kbfs/compare/rpc-cancel .

I'm not sure what the best way to solve this is. @maxtaco any ideas?

**Comment by akalin-keybase** [comment]
Oh, forgot to mention, we should probably also set timeouts on net.Conn used by the Transporter.

**Comment by jzila** [comment]
@akalin-keybase canceling the context will cause the call to return, as long as the transport _isn't already blocked_. Really though, `t.cdec.Write` shouldn't block, it's the read that should block. We could buffer the connection, causing writes to be essentially non-blocking. WDYT?

**Comment by jzila** [comment]
Actually, canceling the context should cause the call to return _right away_ since our more recent changes. See [here](https://github.com/keybase/go-framed-msgpack-rpc/blob/c62326bb4eae20d05bb97a519ae104e1806dde12/dispatch.go#L127) for where the call is caused to return. The subsequent `Encode` can block, but that only affects the communication of the cancel to the server.

**Comment by akalin-keybase** [comment]
Which read should block? As I understand `net.Conn`, `t.cdec.Write` can definitely block. Buffering won't help much -- either we buffer infinitely, or we block once the buffer is full.

Re. your link, the code doesn't even get that far -- it blocks on the Encode call on line 118.

**Comment by jzila** [comment]
OK fair enough, but then the block has nothing to do with the cancellation, and everything to do with the fact that the transport blocked the call. Not sure what the right way is of dealing with that.

**Comment by jzila** [comment]
Copying what I said in slack:
I think it's correct that the transport protocol owns whether the RPC calls block or not. We can control that with timeouts when we establish the connection, then the `Encode` will error correctly.

**Comment by akalin-keybase** [comment]
Okay, after thinking about it, I think I figured it out. You're right that the transport protocol (the `net.Conn`) can choose to block or not block (e.g., if buffering), but we still have to assume that it will block, perhaps indefinitely.

Luckily, we already 'async-wrap' it with writerLoop and encodeCh/encodeResultCh. The problem is that we 'sync-wrap' *that* with the synchronous `Encode()` function. The solution is to make `Encode()` asynchronous and use that. A possible approach (but not the only one) is to:

- Change `encoder` to:

```
type encoder interface {
	Encode(interface{}) <-chan error
}
```

such that `Encode` always returns immediately with a channel on which one can wait for the result.

- Change `handleCall` to:

```
func ... handleCall(...) {
  ...
  v := []interface{}{MethodCall, ...}
  errCh := d.writer.Encode(v)
  select {
    case <-c.ctx.Done():
      // handle cancellation
    case err := <-errCh:
      // handle err
  }
  ...
}
```

As for the `Encode` call for cancellation, we can simply spawn a goroutine that will wait on errCh and, say, log the error if it comes back with one.

- Do something similar everywhere else `Encode` is called.

Once we do this, then cancellation should cause the RPC to immediately return (unless something else is blocking).

**Comment by akalin-keybase** [comment]
One problem this reveals, and which may be worth another issue, is that we have a queue of messages to the server, but cancellation messages are always put at the end, even though it most likely should be put at the head of the queue. We can fix this by something like having `writerLoop` pump two queues, the 'normal' one, and the 'high-priority' one (for cancellations). The tricky part is that there's no built-in way to prioritize channels with select, so we'd have to figure something out there.

**Comment by jzila** [comment]
We can do that pretty easily: you store separate slices for the two queues, and one channel for the queue action. You use the channel to synchronize, but you use the slices for the actual data.

Anyway your solution sounds reasonable, as long as async Encoding is what we actually want. I can see the virtues either way, we just need to make a call.

**Comment by akalin-keybase** [comment]
I'll let @maxtaco weigh in, but I think we do need it -- it was the first thing I ran into when trying to make KBFS take advantage of async RPC cancellation (instead of spawning goroutines itself).

**Comment by maxtaco** [comment]
Just so I have the right high level picture, the setup up:

1. Client calls an RPC to the server
1. Client calls cancel before the server completes via Context
1. Client's RPC library then sends a "cancel" message to the server
1. Server's RPC library then cancels the server hook via Context

**Comment by maxtaco** [comment]
I could be wrong, but if we want to priorities of queues, we might need two connections, since the non-cancelation messages jamming the cancel messages might be queued on the server side.  I'm inclined to think we shouldn't worry about this issue.  We chance really going down a rabbit hole.

**Comment by maxtaco** [comment]
I.e., maybe we should be ok with having the cancel messages stuck in the same queue as the other RPC traffic.

**Comment by jzila** [comment]
@maxtaco the steps you listed describe the happy path. @akalin-keybase wants the cancel to make the call return even if the original call is stuck on its initial `Encode`. A way we can achieve that is to make `Encode` truly asynchronous (which should be easy).

I think your suggestion regarding cancel messages is reasonable, as long as the client realizes that calls they've canceled might still complete.

**Comment by maxtaco** [comment]
:+1: all good!

**Comment by jzila** [comment]
Fixed in #50

**Comment by jzila** [comment]
I started to implement the slice-based solution, and it was insanely hairy. And it still needed synchronization anyway, which either needs to be done with a buffered channel or traditional concurrency (`sync.Cond`).

Here's roughly what I told @maxtaco in Slack:
We still have one loop actually processing the writes, so if a write gets blocked, subsequent accepts to the write queue channel get blocked. A short term solution would be to make that channel buffered, but that just kicks the can down the road. An alternative solution would be to put writes into goroutines, but that has the potential to reorder writes. Another solution would be to use a slice instead of channel to actually store the requests, and traditional concurrency mechanisms like signal/wait on a condition variable to synchronize. However, as I mentioned above, that solution results in very ugly and hairy code.

I'm leaning towards just creating a buffered channel with a buffer size of ~128 (completely arbitrary for now). I've made a PR to that effect in #55

**Comment by maxtaco** [comment]
Ok, I took a look. 

What's wrong with making writes async (putting them into a goroutine?).  I don't see an issue with reordering the writes, as long as the writes are atomic (i.e., all or nothing).

**Comment by jzila** [comment]
@maxtaco Fred and I ended up figuring out a less hairy solution, just by putting `<-ctx.Done()` everywhere.

**Comment by jzila** [comment]
Actually fixed by #55
